smiley/db.py

```diff
@@ -305,104 +336,104 @@ def cache_file_for_run(self, run_id, filename, body):
         else:
             signature_maker.update(body)
         signature = signature_maker.hexdigest()
-        with transaction(self.conn) as c:
-            try:
-                c.execute(
-                    u"""
-                    INSERT INTO file (signature, name, body)
-                    VALUES (:signature, :filename, :body)
-                    """,
-                    {'signature': signature,
-                     'filename': filename,
-                     'body': body,
-                     },
-                )
-            except sqlite3.IntegrityError:
-                pass
-            try:
-                c.execute(
-                    u"""
-                    INSERT INTO run_file
-                    (run_id, signature)
-                    VALUES (:run_id, :signature)
-                    """,
-                    {'run_id': run_id,
-                     'signature': signature,
-                     },
-                )
-            except sqlite3.IntegrityError:
-                pass
-        return signature
-
-    def get_file_signature(self, run_id, filename):
-        """Return the file signature for the named file within the run.
-        """
-        # LOG.debug('get_file_signature(%s)', filename)
-        with transaction(self.conn) as c:
-            c.execute(
+        try:
+            self._insert(
                 u"""
-                SELECT signature
-                FROM file JOIN run_file USING (signature)
-                WHERE
-                  name = :filename
-                  AND
-                  run_id = :run_id
+                INSERT INTO file (signature, name, body)
+                VALUES (:signature, :filename, :body)
                 """,
-                {'filename': filename,
-                 'run_id': run_id,
+                {'signature': signature,
+                 'filename': filename,
+                 'body': body,
                  },
             )
-            row = c.fetchone()
-            # LOG.debug(' -> %s', row)
-            return row['signature'] if row else ''
-
-    def get_files_for_run(self, run_id):
-        with transaction(self.conn) as c:
-            c.execute(
+        except sqlite3.IntegrityError:
+            pass
+        try:
+            # TODO only run this insert if the above one succeeds?
```

Comment:
I imagine if the first insert fails, then this one is guaranteed to throw an IntegrityError.